### DISCUSSION TITLE
Migration failed: Original error - foreign key constraint "constraint_key" cannot be implemented

### DISCUSSION THREAD

**Discussion by Germano123** [mikro-orm/mikro-orm#3428] [Q&A] [answered]
**Describe the bug**
I started a NestJS application from [boileplate](https://github.com/nestjs/typescript-starter.git) and add MikroORM.Everything was just fine until I add the OneToMany relation between School (one) and Supervisor (many). Supervisor and School entities were used in this example.

**Sample code**
```
import { PrimaryKey, Property } from "@mikro-orm/core";

export abstract class AbstractEntity {
    @PrimaryKey({ autoincrement: true })
    id: string;

    @Property()
    createdAt: Date = new Date();

    @Property({ onUpdate: () => new Date() })
    updatedAt: Date = new Date();
}

--------

import { Entity, OneToMany, Property } from "@mikro-orm/core";
import { SupervisorEntity } from "../../../modules/supervisor/entities/supervisor.entity";
import { AbstractEntity } from "../../../common/abstract.entity";

@Entity()
export class SchoolEntity extends AbstractEntity {
    @Property()
    name: string;

    @OneToMany(() => SupervisorEntity, supervisor => supervisor.school, { nullable: true })
    supervisors: SupervisorEntity[];
}
----------

import { Entity, ManyToOne, Property } from "@mikro-orm/core";
import { SchoolEntity } from "../../../modules/school/entities/school.entity";
import { AbstractEntity } from "../../../common/abstract.entity";

@Entity()
export class SupervisorEntity extends AbstractEntity {
    @Property()
    name: string;
    
    @ManyToOne({ entity: () => SchoolEntity, mapToPk: true })
    school: SchoolEntity;
}
```

**Stack trace**
```
Processing 'Migration20220829202913'
MigrationError: Migration Migration20220829202913 (up) failed: Original error: alter table "supervisor_entity" add constraint "supervisor_entity_school_id_foreign" foreign key ("school_id") references "school_entity" ("id") on update cascade; - foreign key constraint "supervisor_entity_school_id_foreign" cannot be implemented
```

**To Reproduce**
1. Create empty project from nestjs boileplate
2. Add mikro orm configuration
3. Add entities to a single OneToMany relation
4. MikroORM migration generate
5. Run app

**Expected behavior**
After the adition of the OneToMany relation it should worked just fine and run the migration up normally.

| Dependency | Version |
| - | - |
| node | 16.16.0 |
| typescript | ^4.7.4 |
| mikro-orm | ^5.3.1 |
| mikro-orm/nestjs | ^5.1.1 |

**Edit:**
Add steps to reproduce.

**Comment by B4nan** [comment, marked as answer]
You have several issues in the entity definition:

- with `mapToPk: true` you map to the PK, not to the entity type
- 1:m relation needs to be represented by `Collection` instance, not array
- collections (1:m or m:n) cannot be nullable
- not sure what that PK is supposed to be doing, autoincrement string?

There are many examples of a m:1/1:m relation, you can check the [example apps](https://github.com/mikro-orm/mikro-orm-examples) or the [ORM tests](https://github.com/mikro-orm/mikro-orm/tree/master/tests). 

My guess is the problem is your PK, you cant have autoincrement string. Try doing this instead:

```ts
@PrimaryKey()
id: number;
```

Or if you want bigint:

```ts
@PrimaryKey({ type: 'bigint' })
id: string;
```

**Reply by Germano123** [reply]
It worked just fine now. I removed the `@PrimaryKey` parameter autoincrement and changed the type from array to `Collection<Entity>`. I tried to add `mapToPk` in `@OneToMany ` decorator but it gives me error but adding in the `@ManyToOne` seems to work.

Why collections can't be nullable? Do u mean that they need to be initialized in the class or they can't be null in DB?

Thx for the quick answer.

**Reply by B4nan** [reply]
> they can't be null in DB?

they dont represent any database column, so yes, they cant be null in db
